libbpfgo/libbpfgo.go

```diff
@@ -478,6 +478,65 @@ func (b *BPFMap) Update(key, value interface{}) error {
 	return nil
 }
 
+type BPFMapIterator struct {
+	b *BPFMap
+
+	err error
+
+	keySize int
+
+	prev []byte
+	next []byte
+}
+
+func (b *BPFMap) Iter(keySize int) *BPFMapIterator {
+	return &BPFMapIterator{
+		b:       b,
+		keySize: keySize,
+		prev:    nil,
+		next:    nil,
+	}
+}
+
+func (it *BPFMapIterator) Next() bool {
+	if it.err != nil {
+		return false
+	}
+
+	prevPtr := unsafe.Pointer(nil)
+	if it.next != nil {
+		prevPtr = unsafe.Pointer(&it.next[0])
+	}
+
+	next := make([]byte, it.keySize)
+	nextPtr := unsafe.Pointer(&next[0])
+
+	errC, err := C.bpf_map_get_next_key(it.b.fd, prevPtr, nextPtr)
+	if errno, ok := err.(syscall.Errno); errC == -1 && ok && errno == C.ENOENT {
+		return false
+	}
+	if err != nil {
+		it.err = err
+		return false
+	}
+
+	it.prev = it.next
+	it.next = next
+
+	return true
+}
+
+// Key returns the current key value of the iterator, if the most recent call to Next returned true.
+// The slice is valid only until the next call to Next.
+func (it *BPFMapIterator) Key() []byte {
+	return it.next
+}
+
+// Err returns the last error that ocurred while table.Iter or iter.Next
+func (it *BPFMapIterator) Err() error {
```

Comment:
Why have this instead of returning it from `Next`?

Reply:
This is the typical iterator pattern in go, this way the usage works like this:

```go
for iterator.Next() {
  // do something with the current value
}
if iterator.Err() != nil {
  // handle error
}
```

Reply:
That makes sense but i'm not convinced that this pattern should be used in an API. It makes it a bit opinionated on how it should be used. I'd rather opt to have as much conformity to libbpf itself as possible.

i.e. I think this would be fine:

```
for err := iterator.Next(); if err == nil { 

} else {
   // handle error
}
```

Reply:
FWIW, I don't think what @brancz proposes is very opinionated. Then the only difference is having to check the boolean (return value of Iterator.Next) before checking the error. However with the idiomatic usage of checking the error at the end, it's nicer to see `for iterator.Next() {`

Reply:
Perhaps it's me that is opinionated then :-) 

To be frank I've always felt that Go shouldn't use any design patterns. Otherwise the whole package has to use the same pattern where applicable and projects that consume it will feel the need to conform to the same pattern. If anything, I think the most idiomatic way/pattern of doing iteration in Go is using callbacks. 

What if we just had `iterator.Next()` return both the key and the error, and got rid of those two methods. Is there any reason why someone would want to call `Next()` and not immediately check both the error and value?

I would also be happy to have a callback parameter to `Next`, in this way we could eliminate having the specify a KeySize when initializing the iterator altogether which further simplifies the API.

Reply:
I just applied what I know and expect from go packages (which could be opinionated due to my experience, but is widely used like this in prometheus tsdb/cockroachdb pebble/various networking libraries eg. [netlink](https://github.com/mdlayher/netlink)).

Reply:
Yea I can't find any example of usage in the standard library so i'd prefer just the simple approach of having `Next()` return the key and the error. If you can make that change I think this is good to go. I'll open up an issue for writing a selftest for this.

I opened up the new libbpfgo repository. I should be able to apply your commit on top of that. If there's any problem with that I'll let you know so that you get the proper credit for it. After that i'll cut the first release.

Reply:
Also thank you very much for this contribution and your patience 😀

Reply:
@grantseltzer I agree that staying as close as possible to libbpf and making thin wrappers is usually preferrable.
In that case, however, I don't think that we loose anything, and the code is still generic (isn't it?).
I also like that `for iterator.Next() {` can be used as @eyakubovich pointed out.

About having a callback to eliminate the need for KeySize, maybe there is another option - using libbpf's `bpf_map__def()` API, we can get the map def:
```
struct bpf_map_def {	
  unsigned int type;	
  unsigned int key_size;	
  unsigned int value_size;	
  unsigned int max_entries;	
  unsigned int map_flags;
};
```

We can even add this to the BPFMap struct, and use it when needed for other places (e.g. for map GetValue() where we also provide the value size)

Reply:
@yanivagman Fair enough. I concede about changing the iterator pattern. @brancz can you please just add a godoc comment above each of the API functions to clarify how it's expected to be used (like what the bool return represents)

Using [`bpf_map_def`](https://github.com/libbpf/libbpf/blob/master/src/libbpf.h#L439) would be fantastic, maybe that should that be part of this PR or have a subsequent change change this, and `GetValue()` as well.

Reply:
Happy to do a follow up with `bpf_map_def`, I'll likely do more for batch operations as well.